long-time operation or many VMs may lead a meta db problem

<!-- Thanks for filing an issue! Before submitting, please fill in the following information. -->

**What happened**
8개 MCIS, VM 100개 이상을 제어하는 상황에서,

MCIS의 개별 VM을 삭제한 후, 

- meta DB의 값이 지워지는 상황 발생. 
- 키는 유지됨.

```
http://localhost:1323/tumblebug/ns/ns01
```
```
Response body: {   "id": "",   "name": "",   "description": "" }
```

동일한 현상이 발생하는 MCIS가 있었음.



**How to reproduce it (as minimally and precisely as possible)**

- 재현해보지 않음
- 유사 환경 조성하여 다양한 테스트 필요
- MCIS에서 특정 VM 삭제 테스트

**Anything else we need to know?**
: 

**Environment**
- Source version or branch: Head
- OS: ubuntu
- Others: 

**Proposed solution**
: 

**Any other context**
:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x47d44b]

goroutine 1033235 [running]:
github.com/cloud-barista/cb-tumblebug/src/core/mcis.GetVmStatusAsync(0xc000e5bc20, 0xc001c6fc12, 0x4, 0xc003a0810e, 0x12, 0xc00184c664, 0x16, 0xc0000dd440, 0x0, 0x0)
	/home/son/go/src/github.com/cloud-barista/cb-tumblebug/src/core/mcis/manageInfo.go:907 +0x28f
created by github.com/cloud-barista/cb-tumblebug/src/core/mcis.GetMcisStatus
	/home/son/go/src/github.com/cloud-barista/cb-tumblebug/src/core/mcis/manageInfo.go:533 +0x6a9
Makefile:12: recipe for target 'run' failed
make: *** [run] Error 2
```

--- a/src/api/grpc/server/mcis/control.go
+++ b/src/api/grpc/server/mcis/control.go
@@ -96,7 +96,7 @@ func (s *MCISService) ControlMcis(ctx context.Context, req *pb.TbMcisActionReque
 
 	logger.Debug("calling MCISService.ControlMcis()")
 
-	result, err := mcis.HandleMcisAction(req.NsId, req.McisId, req.Action)
+	result, err := mcis.HandleMcisAction(req.NsId, req.McisId, req.Action, false)
 	if err != nil {
 		return nil, gc.ConvGrpcStatusErr(err, "", "MCISService.ControlMcis()")
 	}

--- a/src/api/rest/server/mcis/control.go
+++ b/src/api/rest/server/mcis/control.go
@@ -30,7 +30,7 @@ import (
 // @Param nsId path string true "Namespace ID" default(ns01)
 // @Param mcisId path string true "MCIS ID" default(mcis01)
 // @Param action query string true "Action to MCIS" Enums(suspend, resume, reboot, terminate, refine)
-// @Param force query string false "Action to MCIS" Enums(false, true)
+// @Param force query string false "Force control to skip checking controllable status" Enums(false, true)
 // @Success 200 {object} common.SimpleMsg
 // @Failure 404 {object} common.SimpleMsg
 // @Failure 500 {object} common.SimpleMsg

--- a/src/api/rest/server/mcis/control.go
+++ b/src/api/rest/server/mcis/control.go
@@ -30,6 +30,7 @@ import (
 // @Param nsId path string true "Namespace ID" default(ns01)
 // @Param mcisId path string true "MCIS ID" default(mcis01)
 // @Param action query string true "Action to MCIS" Enums(suspend, resume, reboot, terminate, refine)
+// @Param force query string false "Action to MCIS" Enums(false, true)
 // @Success 200 {object} common.SimpleMsg
 // @Failure 404 {object} common.SimpleMsg
 // @Failure 500 {object} common.SimpleMsg
@@ -39,10 +40,15 @@ func RestGetControlMcis(c echo.Context) error {
 	mcisId := c.Param("mcisId")
 
 	action := c.QueryParam("action")
+	force := c.QueryParam("force")
+	forceOption := false
+	if force == "true" {
+		forceOption = true
+	}
 
 	if action == "suspend" || action == "resume" || action == "reboot" || action == "terminate" || action == "refine" {
 
-		result, err := mcis.HandleMcisAction(nsId, mcisId, action)
+		result, err := mcis.HandleMcisAction(nsId, mcisId, action, forceOption)
 		if err != nil {
 			mapA := map[string]string{"message": err.Error()}
 			return c.JSON(http.StatusInternalServerError, &mapA)

--- a/src/api/rest/server/server.go
+++ b/src/api/rest/server/server.go
@@ -82,6 +82,8 @@ func RunServer(port string) {
 	// Middleware
 	e.Use(middleware.Logger())
 	e.Use(middleware.Recover())
+	// limit the application to 20 requests/sec using the default in-memory store
+	e.Use(middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(20)))
 
 	e.HideBanner = true
 	//e.colorer.Printf(banner, e.colorer.Red("v"+Version), e.colorer.Blue(website))
@@ -169,8 +171,13 @@ func RunServer(port string) {
 	e.POST("/tumblebug/mcisDynamicCheckRequest", rest_mcis.RestPostMcisDynamicCheckRequest)
 
 	g.POST("/:nsId/mcisDynamic", rest_mcis.RestPostMcisDynamic)
-	g.GET("/:nsId/mcis/:mcisId", rest_mcis.RestGetMcis)
-	g.GET("/:nsId/mcis", rest_mcis.RestGetAllMcis)
+
+	//g.GET("/:nsId/mcis/:mcisId", rest_mcis.RestGetMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 20 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
+	//g.GET("/:nsId/mcis", rest_mcis.RestGetAllMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 20 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
+	// path specific timeout and ratelimit
+	g.GET("/:nsId/mcis/:mcisId", rest_mcis.RestGetMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 60 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
+	g.GET("/:nsId/mcis", rest_mcis.RestGetAllMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 60 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
+
 	g.PUT("/:nsId/mcis/:mcisId", rest_mcis.RestPutMcis)
 	g.DELETE("/:nsId/mcis/:mcisId", rest_mcis.RestDelMcis)
 	g.DELETE("/:nsId/mcis", rest_mcis.RestDelAllMcis)

--- a/src/api/rest/server/server.go
+++ b/src/api/rest/server/server.go
@@ -175,8 +175,12 @@ func RunServer(port string) {
 	//g.GET("/:nsId/mcis/:mcisId", rest_mcis.RestGetMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 20 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
 	//g.GET("/:nsId/mcis", rest_mcis.RestGetAllMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 20 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
 	// path specific timeout and ratelimit
-	g.GET("/:nsId/mcis/:mcisId", rest_mcis.RestGetMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 60 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
-	g.GET("/:nsId/mcis", rest_mcis.RestGetAllMcis, middleware.TimeoutWithConfig(middleware.TimeoutConfig{Timeout: 60 * time.Second}), middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
+	g.GET("/:nsId/mcis/:mcisId", rest_mcis.RestGetMcis, middleware.TimeoutWithConfig(
+		middleware.TimeoutConfig{Timeout: 60 * time.Second}),
+		middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
+	g.GET("/:nsId/mcis", rest_mcis.RestGetAllMcis, middleware.TimeoutWithConfig(
+		middleware.TimeoutConfig{Timeout: 60 * time.Second}),
+		middleware.RateLimiter(middleware.NewRateLimiterMemoryStore(1)))
 
 	g.PUT("/:nsId/mcis/:mcisId", rest_mcis.RestPutMcis)
 	g.DELETE("/:nsId/mcis/:mcisId", rest_mcis.RestDelMcis)

--- a/src/core/mcis/benchmark.go
+++ b/src/core/mcis/benchmark.go
@@ -160,13 +160,13 @@ func CallMilkyway(wg *sync.WaitGroup, vmList []string, nsId string, mcisId strin
 		common.CBLog.Error(err)
 		errStr = err.Error()
 	}
-	defer res.Body.Close()
 
 	body, err := ioutil.ReadAll(res.Body)
 	if err != nil {
 		common.CBLog.Error(err)
 		errStr = err.Error()
 	}
+	defer res.Body.Close()
 	fmt.Println(string(body))
 
 	fmt.Println("HTTP Status code: " + strconv.Itoa(res.StatusCode))

--- a/src/core/mcis/control.go
+++ b/src/core/mcis/control.go
@@ -54,7 +54,7 @@ type ControlVmResultWrapper struct {
 }
 
 // HandleMcisAction is func to handle actions to MCIS
-func HandleMcisAction(nsId string, mcisId string, action string) (string, error) {
+func HandleMcisAction(nsId string, mcisId string, action string, force bool) (string, error) {
 	action = common.ToLower(action)
 
 	err := common.CheckString(nsId)
@@ -79,7 +79,7 @@ func HandleMcisAction(nsId string, mcisId string, action string) (string, error)
 	if action == "suspend" {
 		fmt.Println("[suspend MCIS]")
 
-		err := ControlMcisAsync(nsId, mcisId, ActionSuspend)
+		err := ControlMcisAsync(nsId, mcisId, ActionSuspend, force)
 		if err != nil {
 			return "", err
 		}
@@ -89,7 +89,7 @@ func HandleMcisAction(nsId string, mcisId string, action string) (string, error)
 	} else if action == "resume" {
 		fmt.Println("[resume MCIS]")
 
-		err := ControlMcisAsync(nsId, mcisId, ActionResume)
+		err := ControlMcisAsync(nsId, mcisId, ActionResume, force)
 		if err != nil {
 			return "", err
 		}
@@ -99,7 +99,7 @@ func HandleMcisAction(nsId string, mcisId string, action string) (string, error)
 	} else if action == "reboot" {
 		fmt.Println("[reboot MCIS]")
 
-		err := ControlMcisAsync(nsId, mcisId, ActionReboot)
+		err := ControlMcisAsync(nsId, mcisId, ActionReboot, force)
 		if err != nil {
 			return "", err
 		}
@@ -119,7 +119,7 @@ func HandleMcisAction(nsId string, mcisId string, action string) (string, error)
 			return "No VM to terminate in the MCIS", nil
 		}
 
-		err = ControlMcisAsync(nsId, mcisId, ActionTerminate)
+		err = ControlMcisAsync(nsId, mcisId, ActionTerminate, force)
 		if err != nil {
 			return "", err
 		}
@@ -252,10 +252,10 @@ func ControlMcis(nsId string, mcisId string, action string) error {
 }
 
 // ControlMcisAsync is func to control MCIS async
-func ControlMcisAsync(nsId string, mcisId string, action string) error {
+func ControlMcisAsync(nsId string, mcisId string, action string, force bool) error {
 
 	checkError := CheckAllowedTransition(nsId, mcisId, action)
-	if checkError != nil {
+	if checkError != nil || !force {
 		return checkError
 	}
 
@@ -462,12 +462,12 @@ func ControlVmAsync(wg *sync.WaitGroup, nsId string, mcisId string, vmId string,
 					common.CBLog.Error(err)
 					return err
 				}
-				defer res.Body.Close()
 				body, err := ioutil.ReadAll(res.Body)
 				if err != nil {
 					common.CBLog.Error(err)
 					return err
 				}
+				defer res.Body.Close()
 
 				fmt.Println("HTTP Status code: " + strconv.Itoa(res.StatusCode))
 				switch {
@@ -678,8 +678,17 @@ func ControlVm(nsId string, mcisId string, vmId string, action string) error {
 		req.Header.Add("Content-Type", "application/json")
 
 		res, err := client.Do(req)
-		defer res.Body.Close()
+		if err != nil {
+			fmt.Println(err)
+			return err
+		}
+
 		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			fmt.Println(err)
+			return err
+		}
+		defer res.Body.Close()
 
 		fmt.Println(string(body))
 
@@ -784,7 +793,7 @@ func CheckAllowedTransition(nsId string, mcisId string, action string) error {
 
 	UpdateMcisInfo(nsId, mcisTmp)
 
-	if strings.Contains(mcisStatusTmp.Status, StatusTerminating) || strings.Contains(mcisStatusTmp.Status, StatusResuming) || strings.Contains(mcisStatusTmp.Status, StatusSuspending) || strings.Contains(mcisStatusTmp.Status, StatusCreating) || strings.Contains(mcisStatusTmp.Status, StatusRebooting) {
+	if strings.Contains(mcisStatusTmp.Status, StatusCreating) || strings.Contains(mcisStatusTmp.Status, StatusTerminating) || strings.Contains(mcisStatusTmp.Status, StatusResuming) || strings.Contains(mcisStatusTmp.Status, StatusSuspending) || strings.Contains(mcisStatusTmp.Status, StatusRebooting) {
 		return errors.New(action + " is not allowed for MCIS under " + mcisStatusTmp.Status)
 	}
 	if !strings.Contains(mcisStatusTmp.Status, "Partial-") && strings.Contains(mcisStatusTmp.Status, StatusTerminated) {

--- a/src/core/mcis/control.go
+++ b/src/core/mcis/control.go
@@ -793,7 +793,12 @@ func CheckAllowedTransition(nsId string, mcisId string, action string) error {
 
 	UpdateMcisInfo(nsId, mcisTmp)
 
-	if strings.Contains(mcisStatusTmp.Status, StatusCreating) || strings.Contains(mcisStatusTmp.Status, StatusTerminating) || strings.Contains(mcisStatusTmp.Status, StatusResuming) || strings.Contains(mcisStatusTmp.Status, StatusSuspending) || strings.Contains(mcisStatusTmp.Status, StatusRebooting) {
+	if strings.Contains(mcisStatusTmp.Status, StatusCreating) ||
+		strings.Contains(mcisStatusTmp.Status, StatusTerminating) ||
+		strings.Contains(mcisStatusTmp.Status, StatusResuming) ||
+		strings.Contains(mcisStatusTmp.Status, StatusSuspending) ||
+		strings.Contains(mcisStatusTmp.Status, StatusRebooting) {
+
 		return errors.New(action + " is not allowed for MCIS under " + mcisStatusTmp.Status)
 	}
 	if !strings.Contains(mcisStatusTmp.Status, "Partial-") && strings.Contains(mcisStatusTmp.Status, StatusTerminated) {

--- a/src/core/mcis/manageInfo.go
+++ b/src/core/mcis/manageInfo.go
@@ -654,21 +654,37 @@ func GetMcisStatus(nsId string, mcisId string) (*McisStatusInfo, error) {
 // GetMcisStatusAll is func to get MCIS status all
 func GetMcisStatusAll(nsId string) ([]McisStatusInfo, error) {
 
-	mcisStatuslist := []McisStatusInfo{}
+	//mcisStatuslist := []McisStatusInfo{}
 	mcisList, err := ListMcisId(nsId)
 	if err != nil {
 		common.CBLog.Error(err)
-		return mcisStatuslist, err
+		return []McisStatusInfo{}, err
 	}
 
+	var wg sync.WaitGroup
+	chanResults := make(chan McisStatusInfo)
+	var mcisStatuslist []McisStatusInfo
+
 	for _, mcisId := range mcisList {
-		mcisStatus, err := GetMcisStatus(nsId, mcisId)
-		if err != nil {
-			common.CBLog.Error(err)
-			return mcisStatuslist, err
-		}
-		mcisStatuslist = append(mcisStatuslist, *mcisStatus)
+		wg.Add(1)
+		go func(nsId string, mcisId string, chanResults chan McisStatusInfo) {
+			defer wg.Done()
+			mcisStatus, err := GetMcisStatus(nsId, mcisId)
+			if err != nil {
+				common.CBLog.Error(err)
+			}
+			chanResults <- *mcisStatus
+		}(nsId, mcisId, chanResults)
 	}
+
+	go func() {
+		wg.Wait()
+		close(chanResults)
+	}()
+	for result := range chanResults {
+		mcisStatuslist = append(mcisStatuslist, result)
+	}
+
 	return mcisStatuslist, nil
 
 	//need to change status
@@ -775,6 +791,7 @@ func GetVmCurrentPublicIp(nsId string, mcisId string, vmId string) (TbVmStatusIn
 				return http.ErrUseLastResponse
 			},
 		}
+		defer client.CloseIdleConnections()
 		req, err := http.NewRequest(method, url, strings.NewReader(string(payload)))
 
 		errorInfo.Status = StatusFailed
@@ -984,7 +1001,9 @@ func GetVmStatus(nsId string, mcisId string, vmId string) (TbVmStatusInfo, error
 				CheckRedirect: func(req *http.Request, via []*http.Request) error {
 					return http.ErrUseLastResponse
 				},
+				Timeout: 30 * time.Second,
 			}
+			defer client.CloseIdleConnections()
 
 			// Retry to get right VM status from cb-spider. Sometimes cb-spider returns not approriate status.
 			retrycheck := 2

--- a/src/core/mcis/manageInfo.go
+++ b/src/core/mcis/manageInfo.go
@@ -791,7 +791,6 @@ func GetVmCurrentPublicIp(nsId string, mcisId string, vmId string) (TbVmStatusIn
 				return http.ErrUseLastResponse
 			},
 		}
-		defer client.CloseIdleConnections()
 		req, err := http.NewRequest(method, url, strings.NewReader(string(payload)))
 
 		errorInfo.Status = StatusFailed
@@ -809,8 +808,12 @@ func GetVmCurrentPublicIp(nsId string, mcisId string, vmId string) (TbVmStatusIn
 			return errorInfo, err
 		}
 
-		defer res.Body.Close()
 		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			fmt.Println(err)
+			return errorInfo, err
+		}
+		defer res.Body.Close()
 
 		statusResponseTmp = statusResponse{}
 
@@ -1001,12 +1004,11 @@ func GetVmStatus(nsId string, mcisId string, vmId string) (TbVmStatusInfo, error
 				CheckRedirect: func(req *http.Request, via []*http.Request) error {
 					return http.ErrUseLastResponse
 				},
-				Timeout: 30 * time.Second,
+				Timeout: 60 * time.Second,
 			}
-			defer client.CloseIdleConnections()
 
 			// Retry to get right VM status from cb-spider. Sometimes cb-spider returns not approriate status.
-			retrycheck := 2
+			retrycheck := 5
 			for i := 0; i < retrycheck; i++ {
 
 				req, err := http.NewRequest(method, url, strings.NewReader(string(payload)))
@@ -1041,7 +1043,7 @@ func GetVmStatus(nsId string, mcisId string, vmId string) (TbVmStatusInfo, error
 				if statusResponseTmp.Status != "" {
 					break
 				}
-				time.Sleep(1 * time.Second)
+				time.Sleep(5 * time.Second)
 			}
 
 		} else {

--- a/src/core/mcis/manageInfo.go
+++ b/src/core/mcis/manageInfo.go
@@ -1368,14 +1368,14 @@ func DelMcis(nsId string, mcisId string, option string) error {
 		if strings.EqualFold(option, ActionTerminate) {
 
 			// ActionRefine
-			_, err := HandleMcisAction(nsId, mcisId, ActionRefine)
+			_, err := HandleMcisAction(nsId, mcisId, ActionRefine, true)
 			if err != nil {
 				common.CBLog.Error(err)
 				return err
 			}
 
 			// ActionTerminate
-			_, err = HandleMcisAction(nsId, mcisId, ActionTerminate)
+			_, err = HandleMcisAction(nsId, mcisId, ActionTerminate, true)
 			if err != nil {
 				common.CBLog.Error(err)
 				return err

--- a/src/core/mcis/monitoring.go
+++ b/src/core/mcis/monitoring.go
@@ -132,13 +132,13 @@ func CheckDragonflyEndpoint() error {
 			fmt.Println(err)
 			return err
 		}
-		defer res.Body.Close()
 
 		body, err := ioutil.ReadAll(res.Body)
 		if err != nil {
 			fmt.Println(err)
 			return err
 		}
+		defer res.Body.Close()
 
 		fmt.Println(string(body))
 		return nil
@@ -256,15 +256,16 @@ func CallMonitoringAsync(wg *sync.WaitGroup, nsID string, mcisID string, mcisSer
 			errStr += "/ " + err.Error()
 		}
 
-		defer res.Body.Close()
 		body, err := ioutil.ReadAll(res.Body)
 		if err != nil {
 			common.CBLog.Error(err)
 			errStr += "/ " + err.Error()
 		}
+		defer res.Body.Close()
 
 		result = string(body)
 	}
+	defer res.Body.Close()
 
 	//wg.Done() //goroutin sync done
 
@@ -479,14 +480,15 @@ func CallGetMonitoringAsync(wg *sync.WaitGroup, nsID string, mcisID string, vmID
 				errStr = err1.Error()
 			}
 
-			defer res.Body.Close()
 			body, err2 := ioutil.ReadAll(res.Body)
 			if err2 != nil {
 				common.CBLog.Error(err2)
 				errStr = err2.Error()
 			}
+			defer res.Body.Close()
 			response = string(body)
 		}
+
 	} else {
 		reqParams := df_pb.VMOnDemandMonQryRequest{
 			NsId:    nsID,

--- a/src/core/mcis/provisioning.go
+++ b/src/core/mcis/provisioning.go
@@ -1470,13 +1470,12 @@ func CreateVm(nsId string, mcisId string, vmInfoData *TbVmInfo, option string) e
 			return err
 		}
 
-		defer res.Body.Close()
 		body, err := ioutil.ReadAll(res.Body)
-
 		if err != nil {
 			common.CBLog.Error(err)
 			return err
 		}
+		defer res.Body.Close()
 
 		tempSpiderVMInfo = SpiderVMInfo{} // FYI; SpiderVMInfo: the struct in CB-Spider
 		err = json.Unmarshal(body, &tempSpiderVMInfo)

